app/config.go

```diff
@@ -208,6 +209,30 @@ func (a *App) ensureAsymmetricSigningKey() error {
 	return nil
 }
 
+func (a *App) ensureInstallationDate() error {
+	_, err := a.getSystemInstallDate()
+	if err == nil {
+		return nil
+	}
+
+	result := <-a.Srv.Store.User().InferSystemInstallDate()
+	var installationDate int64
+	if result.Err == nil && result.Data.(int64) > 0 {
+		installationDate = result.Data.(int64)
+	} else {
+		installationDate = utils.MillisFromTime(time.Now())
+	}
+
+	result = <-a.Srv.Store.System().SaveOrUpdate(&model.System{
```

Comment:
@saturninoabril here. This method try to get the system install date (using the InstallationDate entry in the Systems table) if it can, return the date and that's it, but if it can't try to infer the date (based on the users existing in the system, maybe the system has no users yet), so It will use the current date. Once you have the installation date (inferred or using "now"), It saves the date in the Systems Table. The next time, it will pick the date directly from the Systems Table.

Reply:
Right, I've seen it.  Does this change covers saving of version and installation date during system upgrade?

Reply:
I think it covers almost any case. If you have previous version installed, and you upgrade, the first time you run the new "ensure" code will try to infer you installation date based on your current existing users (the oldest one, will be your installation date). In the case that you are installing a fresh install, It won't find any user, so, it will use the current date.

There are weird cases where you upgrade, and for any reason, you can't get any user (the connection to the server fails, for example) but you can write the system table milliseconds later. This is very weird scenario.

Other case would be a user that have the mattermost server installed for X time, but have no users at all. If this happens, after the upgrade, the installation date will be the current date. But again a very weird scenario.

Reply:
Cool, thanks for the explanation.  I asked because we have this similar ticket for Net Promoter Score (NPS) - https://mattermost.atlassian.net/browse/MM-11457

Reply:
@saturninoabril Yes, for now we are storing only the installation date. In the future would be nice to have the "History" of the upgrades. we can follow the same approach (something like "ensure full-history") and have a "VersionHistory" entry in the Systems Table, and store there a json with something like:
```json
{
  "5.1.0": 1533552194000
  "5.2.0": 1533564189414
}
```

And them check at the beginning, "Is the current version in the VersionHistory?, Yes, then nothing to do, No, then add it with the current time."